### DISCUSSION TITLE
Prisma Object with string key type

### DISCUSSION THREAD

**Discussion by seogrady** [prisma/prisma#13989] [Q&A] [answered]
I want to store an Object in MongoDB using Prisma and I want to store as a nested Dict (same collection) rather than as an Array, but still maintain types.

So...
`{ [id: string ]: Object }`

Rather than...
`[Object,Object]`

Is this possible?

**Comment by TasinIshmam** [comment]
Hey @seogrady 👋  

Perhaps you're looking for [composite types](https://www.prisma.io/docs/concepts/components/prisma-client/composite-types)? They allow you to embed an array of objects inside a MongoDB document.

**Reply by seogrady** [reply]
Hi @TasinIshmam – Yes, but it's not possible to have a String type as an object key, correct? For example, storing data like a hash table. I'm gathering you have to use JSON type.

**Reply by TasinIshmam** [reply]
Are you trying to store _unstrucutred_ data without a particular schema? In that case, you should use the JSON type or JSON array type (`Json[]`). If your data has a specific structure, you should use the composite types. Also, all object keys (what Prisma calls `fields`) are basically serialized to string when you're interacting with it, so not totally sure what you mean. 

Again, you can mix and match as well, your composite type can have JSON fields for the unstructured portions. 

Perhaps you could give an example with concrete code? That might help me understand the problem a bit better.

**Reply by seogrady** [reply]
@TasinIshmam – Sure thing!

In typescript you can have...
```
type Test = {
  [id: string]: Object;
};

const test: Test = { 'id1': {} }
```

Can you have this in Prisma? I'm guessing you'd need to use JSON type.

It would be nice if you could use this with Prisma and maintain type structure. For example, something like...

```
type Test {
  [String] SomeCompositeType
}
```

**Reply by TasinIshmam** [reply]
Oh okay, I think composite types and JSON fields are exactly what you're looking for. Composite types will maintain type structures (Prisma will automatically generate the types for you). JSON fields won't have types as it's just arbitrary JSON. 

If your schema is well defined, composite types are the way to go.

**Reply by janpio** [reply]
Can you share two different example data entries of this @seogrady? If I understand correctly, composite types might actually not be the way to go as they required all objects to have the same key names - not just the type of the key.

**Reply by seogrady** [reply]
You're right @janpio. It seems Prisma is strict with key names and they are required to be defined.

Perhaps in the future this could be changed to make the schema more flexible. There's no reason you couldn't have a Record/Dictionary type if you allow Array types.

**Reply by TasinIshmam** [reply]
Ah okay, I think I understand your problem now. Mongoose [supports something like this](https://stackoverflow.com/a/12389168/8855844), but Prisma cannot have unspecified/dynamic fields at the model level at the moment. 

Would be great if you could create a new [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) for this so we can track interest in this feature @seogrady.

**Reply by TasinIshmam** [reply]
Hi @seogrady, 

I created [an issue](https://github.com/prisma/prisma/issues/14481) about adding this feature to Prisma, please feel free to leave a 👍  or comment your use-case around this over there.

**Comment by janpio** [comment, marked as answer]
Knowing the field names of a model is one of the basic concepts of Prisma. It allows us to generate TypeScript types that can tell you waht type a specific field is expected to be and similar.

But you can still open a feature request for this. Maybe many others are interested in it, and maybe there is a type safe way to provider such an API.

**Reply by seogrady** [reply]
Thanks @janpio! You should still be able to provide a type-safe way to do this. The object key would be a unique string and the value would be a strongly typed object, array or primitive. I'll open a feature request 👍

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
